### DISCUSSION TITLE
Support underlays in modern more

### DISCUSSION THREAD

**Discussion by maxrjones** [GenericMappingTools/gmt#4995] [Feature development discussion]
**Description of the desired feature**

Add the functionality to recycle bits of PostScript in modern mode, e.g., make a common underlay base.ps and use it in different plots. I'm most interested in the basemap use case, but there also may be people who are interested in being able to add common overlays to many plots.

A pseudo-code example:

```
export GMT_SESSION_NAME=underlay
gmt begin
gmt basemap -JM6i -R-30/30/-30/30 -Bafg

unset GMT_SESSION_NAME

gmt begin -Dunderlay,overlay # duplicate underlay session to overlay
gmt plot ex1.txt -S -Cchrome
gmt end show

gmt begin -Dunderlay,overlay # duplicate underlay session to overlay
gmt plot ex2.txt -S -Cchrome
gmt end show

gmt begin -Dunderlay,overlay # duplicate underlay session to overlay
gmt plot ex3.txt -S -Cchrome
gmt end show
```

**Are you willing to help implement and maintain this feature?** Yes/No

**Comment by PaulWessel** [comment]
Some of you may recall an early modern mode command called revert (revert.c now lives in exile in the sandbox just in case).  It could strip off layers from the current plot.  It seems a bit clutzy to me thought.

Here are quick thoughts on overlays/underlays

1. Plot what you consider a layer normally
2. Run **gmt layer get** _name_ to make a named snapshot of current plot
3. When you want to add this layer to another plot, run **gmt layer put** _name_
4. To use it as an underlay (first layer) for a new figure, you just do **gmt figure** first then **gmt layer put** _name_
5. Layers are not tied to a figure, but tied to session, and may be used any number of times
6. Named layers die when the session ends

The verbs **get** and **put** can be discussed.  We can discuss if calling get with the same name as before simply overwrites or if we refuse to do so without some **-F** force option or an intermediate **gmt layer delete** _name_ command.

**Comment by maxrjones** [comment]
> Some of you may recall an early modern mode command called revert (revert.c now lives in exile in the sandbox just in case). It could strip off layers from the current plot. It seems a bit clutzy to me thought.
> 
> Here are quick thoughts on overlays/underlays
> 
>     1. Plot what you consider a layer normally
> 
>     2. Run **gmt layer get** _name_ to make a named snapshot of current plot
> 
>     3. When you want to add this layer to another plot, run **gmt layer put** _name_
> 
>     4. To use it as an underlay (first layer) for a new figure, you just do **gmt figure** first then **gmt layer put** _name_
> 
>     5. Layers are not tied to a figure, but tied to session, and may be used any number of times
> 
>     6. Named layers die when the session ends
> 
> 
> The verbs **get** and **put** can be discussed. We can discuss if calling get with the same name as before simply overwrites or if we refuse to do so without some **-F** force option or an intermediate **gmt layer delete** _name_ command.

Given that the layers die when the session ends, it is not clear to me how you would finalize the figures without closing the session if, for example, you want to use the same layers when updating a plot every day, week, year, etc.

**Comment by PaulWessel** [comment]
Perhaps we should convert this to a discussion item?

**Reply by maxrjones** [reply]
A discussion item, now

**Reply by PaulWessel** [reply]
If we let layers outlive the session then they must be copied to the current directory.  Now the user becomes the custodian of half-baked PostScript files and in charge of juggling these.  I understand your argument but it becomes a question of whether or not we want modern mode to allow for user meddling in files or we we just say that such very special processing cases should be done in classic mode.  One reason for designing modern mode was to keep it much simpler and we acknowledged that some tasks that can be done in classic mode cannot be done (the same way) in modern mode.  The beauty of modern mode is the containment of all intermediate files in the session directory until completion, then removal of all such files, and the blissful ignorance of PostScript files.  This proposal clearly would leave layer files around that cannot be viewed.  I guess I am unconvinced that the vast majority of GMT users need this capability.  I think it would be easy to implement, but...

**Reply by maxrjones** [reply]
I don't see why they must be copied to the current directory. I like my original idea of leaving a session open and then having an option to duplicate it into a new session. The set session name is very useful but I doubt many people are aware that it could be more than just a error-fixing tool for getting the process ID. This duplication of session option would mean that everything is still contained in the .session directory, but allows for more flexibility for the small subset of GMT users who want this capability without wanting to move backwards to classic mode (which would never allow underlays + subplots, for example).

**Reply by KristofKoch** [reply]
This sounds to me like classic mode with a modern mode veneer. Why reinvent classic mode? I agree with @PaulWessel on this one. And while classic mode surely has its pitfalls (and a learning curve way steeper than that for modern mode), it gives users way more power on what can be accomplished. 

I might be a niche user – but this fine grained control over the plot generation process of classic mode (then the *only* mode) was one of the key points that drove me to GMT. I see it a bit like MacOS: Modern mode is like the polished MacOS GUI which allows you to accomplish most of everyday tasks with ease. But if you need very fine grain control or raw power (and responsibilities that come with it) you are always able unleash the beast in form of the Unix system beneath the GUI (classic mode in our GMT universe). Why not keep that separation?

**Reply by maxrjones** [reply]
Good points @KristofKoch. I will accept that I will need to keep using -K's and -O's for the most powerful map control features in GMT :smile:

**Reply by seisman** [reply]
I think we should have a section explaining how classic mode works in the documentation. For new GMT users, they only know modern mode, and know nothing about -K, -O and PS layers unless they read the old documentation.

**Reply by PaulWessel** [reply]
Yes, there may need to be a high-level overview of the two systems, the pros and cons, what can and cannot be done in modern and what is awkward in classic, etc.

**Reply by anbj** [reply]
People should be told that the (potential) convenience and simplicity of modern mode has a price.
I've tried modern mode a couple of times, but I keep falling back on classic mode; it's what I've been using since I first touched GMT, and the _fingerspitzengefühl_ it gives you, is of high value to me. Also I'm getting grumpier by the year. I'm glad I learned GMT when only classical mode existed. I've learned things that I probably never would, if only modern mode was around.

**Reply by maxrjones** [reply]
One option would be to revise this section in the cookbook (https://docs.generic-mapping-tools.org/dev/cookbook/introduction.html#modern-and-classic-mode) to include the high-level overview, pros and cons, etc. for each without assuming any prior knowledge about GMT.

**Comment by joa-quim** [comment]
Modern forbids the use of ``-O, -K, >, >>`` so one cannot play with pieces of a PS file. But other that, what else is not possible in Modern but yes in Classic?
